File: problems/039-calibration-multiple-choice-knapsack/solutions/python/main.py
import sys

t = list(map(int, sys.stdin.buffer.read().split()))
it = iter(t)
G, C = next(it), next(it)
dp = [0] * (C + 1)
for _ in range(G):
    k = next(it)
    a = [(next(it), next(it)) for _ in range(k)]
    ndp = dp[:]
    for w, v in a:
        for c in range(w, C + 1):
            ndp[c] = max(ndp[c], dp[c - w] + v)
    dp = ndp
print(dp[C])

File: problems/039-calibration-multiple-choice-knapsack/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let g: usize = t.next().unwrap().parse().unwrap();
    let cmax: usize = t.next().unwrap().parse().unwrap();
    let mut dp = vec![0u64; cmax + 1];
    for _ in 0..g {
        let k: usize = t.next().unwrap().parse().unwrap();
        let mut a = Vec::with_capacity(k);
        for _ in 0..k {
            a.push((
                t.next().unwrap().parse::<usize>().unwrap(),
                t.next().unwrap().parse::<u64>().unwrap(),
            ));
        }
        let mut next = dp.clone();
        for (w, v) in a {
            for c in w..=cmax {
                next[c] = next[c].max(dp[c - w] + v);
            }
        }
        dp = next;
    }
    println!("{}", dp[cmax]);
}
